### PR TITLE
(#64) Add segments-progress endpoint to the dashboard-api

### PR DESCRIPTION
**Pull request by ericrfdev**

### CODE REVIEW

File: dashboard-api/internal/platform/server/handler/segmentsprogress/list_test.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+type response struct {
+	StatusCode int
+	Header     http.Header
+	body       string
+}
+
+func performRequest(path string, studyRepository studiesmanager.StudyRepository, studySegmentsRepository studiesmanager.StudySegmentsRepository) response {
+	gin.SetMode(gin.TestMode)
+	r := gin.New()
+	r.GET("/studies/:slug/segments-progress", ListHandler(studyRepository, studySegmentsRepository))
```
Comment (nandanrao) [inline dashboard-api/internal/platform/server/handler/segmentsprogress/list_test.go:138]: Why is the path redefined in the tests?

Comment (nandanrao) [reply dashboard-api/internal/platform/server/handler/segmentsprogress/list_test.go:138]: Your `Server` struct anyways is given the repositories and has a `RegisterRoutes`. Why don't you use that?

File: dashboard-api/internal/platform/storage/cockroachdb/study_segments_repository.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	rows, err := r.db.Query("SELECT created, details FROM optimization_reports WHERE report_type = 'FACEBOOK_OPTIMIZATION' and study_id = $1 ORDER BY created ASC", studyId)
+	if err != nil {
+		return nil, fmt.Errorf("(db.Query) error trying to get all time segments progress (studyId: %s): %v", studyId, err)
+	}
+
+	defer rows.Close()
+
+	for rows.Next() {
+		var created time.Time
+		var detailsInBytes []byte
```
Comment (nandanrao) [inline dashboard-api/internal/platform/storage/cockroachdb/study_segments_repository.go:47]: pgx allows you to scan json directly to structs.

File: dashboard-api/internal/platform/storage/cockroachdb/study_segments_repository_test.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+		"SELECT created, details FROM optimization_reports WHERE report_type = 'FACEBOOK_OPTIMIZATION' and study_id = $1 ORDER BY created ASC").
+		WithArgs(studyId).
+		WillReturnRows(rows)
+
+	repo := NewStudySegmentsRepository(db)
+
+	_, err = repo.GetAllTimeSegmentsProgress(context.Background(), studyId)
+
+	assert.NoError(t, sqlMock.ExpectationsWereMet())
+	assert.Equal(t, errors.New("(rows.Scan) error trying to get all time segments progress (studyId: c463c577-a3d1-482e-b638-10e6733e2325); sql: expected 1 destination arguments in Scan, not 2"), err)
```
Comment (nandanrao) [inline dashboard-api/internal/platform/storage/cockroachdb/study_segments_repository_test.go:52]: I don't understand what we're testing here. Is this a realistic runtime error?

File: dashboard-api/internal/platform/storage/cockroachdb/study_segments_repository_test.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+	studiesmanager "github.com/vlab-research/vlab/dashboard-api/internal"
+)
+
+func Test_StudySegmentsRepository_GetAllTimeSegmentsProgress_UnexpectedError_during_Query(t *testing.T) {
+	studyId := "c463c577-a3d1-482e-b638-10e6733e2325"
+	db, sqlMock, err := sqlmock.New(sqlmock.QueryMatcherOption(sqlmock.QueryMatcherEqual))
+	require.NoError(t, err)
+
+	sqlMock.ExpectQuery(
+		"SELECT created, details FROM optimization_reports WHERE report_type = 'FACEBOOK_OPTIMIZATION' and study_id = $1 ORDER BY created ASC").
```
Comment (nandanrao) [inline dashboard-api/internal/platform/storage/cockroachdb/study_segments_repository_test.go:22]: I'm not sure I see the advantage of this sql mock library. Either I want to have integration tests to test everything with the database included, or I want unit tests where I mock that we can handle edge cases. Why do I care what the query is if I anyways want to test the query does what I want? It's brittle, you have to change the tests when the query changes, even if the rest of your business logic and desired behavior is the same.

**Review by nandanrao — COMMENTED** [review]
Left some comments - overall things look really good all things considered. But I still feel like things are heavier and more brittle than they need to be.

Do we have integration tests with the database included right now? Or no?

File: dashboard-api/internal/platform/server/handler/segmentsprogress/list_test.go (hunk 1 above)
Comment (ericrfdev) [reply dashboard-api/internal/platform/server/handler/segmentsprogress/list_test.go]: Refactored the code to take advantage of it 👌

**Comment by ericrfdev** [comment]
@nandanrao Added two new commits to handle the things we discussed in our previous meeting:
* Centralize bootstraping  logic of the repositories inside of the storage package and refactor server and handlers to receive all the repositories
* Create a helper to perform get request during tests and refactor existing handler tests to use it

When we have the app dockerized, I'll review existing test and refactor all the happy path tests to be integration one. In the process if I find out some tests that are not longer relevant I'll remove them too 👌

**Comment by nandanrao** [comment]
LGTM! Thanks for making those changes, looks much more straight-forward to me now.
